File: sdk/ios/Sources/GISOTracker/ExposureTracker.swift
import Foundation
import UIKit

/// 曝光监测：CADisplayLink 低频轮询可视比例。
/// 口径与 Android/Web 一致：可视面积 ≥ ratio 持续 ≥ duration 记一次；
/// 滚出（可视 < 20%）后可重记；单次页面进入内每实例最多 maxPerPage 次。
final class ExposureTracker {
    private static let exitRatio = 0.2
    private static let pollInterval: TimeInterval = 0.25

    private var ratio: Double
    private var duration: TimeInterval
    private var maxPerPage: Int
    private let onExposure: (UIView, Int, Double) -> Void  // (view, durMs, maxRatio)

    private final class Pending {
        var enterTs = Date()
        var maxRatio = 0.0
    }

    private var observed = NSHashTable<UIView>.weakObjects()
    private var pending = [ObjectIdentifier: Pending]()
    private var counts = [ObjectIdentifier: Int]()
    private var timer: Timer?

    init(ratio: Double, duration: TimeInterval, maxPerPage: Int,
         onExposure: @escaping (UIView, Int, Double) -> Void) {
        self.ratio = ratio
        self.duration = duration
        self.maxPerPage = maxPerPage
        self.onExposure = onExposure
    }

    func observe(_ view: UIView) {
        observed.add(view)
        startTimerIfNeeded()
    }

    /// 远程配置下发后更新口径
    func updateThresholds(ratio: Double, duration: TimeInterval, maxPerPage: Int) {
        self.ratio = ratio
        self.duration = duration
        self.maxPerPage = maxPerPage
    }

    /// 页面切换时调用：重置每页曝光计数
    func resetPage() {
        counts.removeAll()
        pending.removeAll()
    }

    private func startTimerIfNeeded() {
        guard timer == nil else { return }
        timer = Timer.scheduledTimer(withTimeInterval: Self.pollInterval, repeats: true) { [weak self] _ in
            self?.poll()
        }
    }

    private func poll() {
        for view in observed.allObjects {
            let id = ObjectIdentifier(view)
            let r = Self.visibleRatio(view)
            if r >= ratio {
                if let p = pending[id] {
                    p.maxRatio = max(p.maxRatio, r)
                    if Date().timeIntervalSince(p.enterTs) >= duration {
                        pending.removeValue(forKey: id)
                        let count = counts[id] ?? 0
                        if count < maxPerPage {
                            counts[id] = count + 1
                            let durMs = Int(Date().timeIntervalSince(p.enterTs) * 1000)
                            onExposure(view, durMs, (p.maxRatio * 100).rounded() / 100)
                        }
                    }
                } else {
                    let p = Pending()
                    p.maxRatio = r
                    pending[id] = p
                }
            } else if r < Self.exitRatio {
                pending.removeValue(forKey: id)
            }
        }
    }

    private static func visibleRatio(_ view: UIView) -> Double {
        guard let window = view.window, !view.isHidden, view.alpha > 0.01 else { return 0 }
        let frame = view.convert(view.bounds, to: window)
        let visible = frame.intersection(window.bounds)
        guard !visible.isNull, frame.width > 0, frame.height > 0 else { return 0 }
        return Double((visible.width * visible.height) / (frame.width * frame.height))
    }
}

File: sdk/android/src/main/java/com/giso/tracker/ExposureTracker.java
package com.giso.tracker;

import android.graphics.Rect;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.view.ViewTreeObserver;

import java.util.Map;
import java.util.WeakHashMap;

/**
 * 曝光监测：基于 ViewTreeObserver 的 onDraw/onScrollChanged 轮询可视比例。
 * 口径与 Web 端一致：可视面积 ≥ ratio 持续 ≥ duration 记一次；
 * 滚出（可视 < 20%）后可重记；单次页面进入内每实例最多 maxPerPage 次。
 */
final class ExposureTracker {
    interface Listener {
        void onExposure(View view, long durationMs, float maxRatio);
    }

    private static final float EXIT_RATIO = 0.2f;

    private volatile float ratio;
    private volatile long durationMs;
    private volatile int maxPerPage;
    private final Listener listener;
    private final Handler handler = new Handler(Looper.getMainLooper());

    private final WeakHashMap<View, Pending> pending = new WeakHashMap<>();
    private WeakHashMap<View, Integer> counts = new WeakHashMap<>();
    private final WeakHashMap<View, Boolean> observed = new WeakHashMap<>();

    private static final class Pending {
        long enterTs;
        float maxRatio;
        Runnable timer;
    }

    ExposureTracker(float ratio, long durationMs, int maxPerPage, Listener listener) {
        this.ratio = ratio;
        this.durationMs = durationMs;
        this.maxPerPage = maxPerPage;
        this.listener = listener;
    }

    void observe(final View view) {
        if (observed.containsKey(view)) return;
        observed.put(view, Boolean.TRUE);
        ViewTreeObserver vto = view.getViewTreeObserver();
        ViewTreeObserver.OnPreDrawListener l = () -> {
            check(view);
            return true;
        };
        vto.addOnPreDrawListener(l);
        view.addOnAttachStateChangeListener(new View.OnAttachStateChangeListener() {
            @Override public void onViewAttachedToWindow(View v) { }
            @Override public void onViewDetachedFromWindow(View v) {
                clearPending(v);
                if (v.getViewTreeObserver().isAlive()) {
                    v.getViewTreeObserver().removeOnPreDrawListener(l);
                }
                observed.remove(v);
            }
        });
    }

    /** 远程配置下发后更新口径 */
    void updateThresholds(float ratio, long durationMs, int maxPerPage) {
        this.ratio = ratio;
        this.durationMs = durationMs;
        this.maxPerPage = maxPerPage;
    }

    /** 页面切换时调用：重置每页曝光计数 */
    void resetPage() {
        counts = new WeakHashMap<>();
        for (Map.Entry<View, Pending> e : pending.entrySet()) {
            if (e.getValue().timer != null) handler.removeCallbacks(e.getValue().timer);
        }
        pending.clear();
    }

    private void check(View view) {
        float r = visibleRatio(view);
        Pending p = pending.get(view);
        if (r >= ratio) {
            if (p == null) {
                final Pending np = new Pending();
                np.enterTs = System.currentTimeMillis();
                np.maxRatio = r;
                np.timer = () -> fire(view);
                pending.put(view, np);
                handler.postDelayed(np.timer, durationMs);
            } else {
                p.maxRatio = Math.max(p.maxRatio, r);
            }
        } else if (r < EXIT_RATIO && p != null) {
            clearPending(view);
        }
    }

    private void fire(View view) {
        Pending p = pending.remove(view);
        if (p == null) return;
        Integer c = counts.get(view);
        int count = c == null ? 0 : c;
        if (count >= maxPerPage) return;
        counts.put(view, count + 1);
        listener.onExposure(view, System.currentTimeMillis() - p.enterTs,
                Math.round(p.maxRatio * 100f) / 100f);
    }

    private void clearPending(View view) {
        Pending p = pending.remove(view);
        if (p != null && p.timer != null) handler.removeCallbacks(p.timer);
    }

    private static float visibleRatio(View view) {
        if (!view.isShown() || view.getWidth() == 0 || view.getHeight() == 0) return 0f;
        Rect visible = new Rect();
        if (!view.getGlobalVisibleRect(visible)) return 0f;
        float visibleArea = visible.width() * (float) visible.height();
        float totalArea = view.getWidth() * (float) view.getHeight();
        return totalArea <= 0 ? 0f : visibleArea / totalArea;
    }
}
